### PR TITLE
Add guard word before local var CMiniColDef[9]

### LINKED ISSUE #73718: illegal read of byte preceding an automatic (stack allocate) variable
### Description

Compiling on/for x64 architecture on linux with clang-14.

The function CMiniMdBase::InitColsForTable

https://github.com/dotnet/runtime/blob/9bd2cb52375ee073ffbe24da1c425788bc88704f/src/coreclr/md/runtime/metamodel.cpp#L720
allocates an automatic array variable of CMiniColDef[9].  The base address of this array is passed down through UsesAllocatedMemory().  UsesAllocatedMemory apparently implements some kind of marking of the *preceding* byte to determine if the block was obtained from a memory allocator.

But in this case, the block is an automatic in some nearby frame.  In our case, the preceding bytes contain a pointer pTemplate.  Uses AllocatedMemory reads a byte from that pointer, which one depends on endianness.

You can't do thatt.

### Reproduction Steps

Compile with -fsanitize=address and stop at first fail.

### Expected behavior

no errors detected by asan

### Actual behavior

Read an adjacent memory location.

### Regression?

_No response_

### Known Workarounds

Probably put a dummy slot in the frame before the [9] vector to emulate what's done when those CMiniColDefs are allocated dynamically.

### Configuration

_No response_

### Other information

_No response_

### PR DESCRIPTION
**Pull request by RobertHenry6bev**
Initialize the guard word so it does not pass the UsesAllocatedMemory
test.

Found running output of clang14 -fsanitize=address, then inspection

Fixes https://github.com/dotnet/runtime/issues/73718

### CODE REVIEW

**Review by AntonLapounov — CHANGES_REQUESTED** [review]
Please use something like below for consistency with other places:
```C++
    const int MaxCols = 9;
    BYTE tempCols[MaxCols * sizeof(CMiniColDef) + 1];

    _ASSERTE(MaxCols >= pTable->m_cCols);
    // Mark the array of columns as not allocated (not ALLOCATED_MEMORY_MARKER) for SetNewColumnDefinition call bellow
    tempCols[0] = 0;
    CMiniColDef *pCols = BYTEARRAY_TO_COLDES(tempCols);
```

File: src/coreclr/md/runtime/metamodel.cpp [hunk 1]
```diff
@@ -717,7 +717,17 @@ CMiniMdBase::InitColsForTable(
                                     // should we write the data into the structure
 {
     const CMiniTableDef *pTemplate;     // Template table definition.
-    CMiniColDef pCols[9];               // The col defs to init.
+    const int MaxCols = 9;
+    typedef uint32_t markword_t;
+    BYTE tempCols[sizeof(markword_t) + MaxCols * sizeof(CMiniColDef)]; // keep aligned
```
Comment (AntonLapounov) [inline src/coreclr/md/runtime/metamodel.cpp:722]: We use a single-byte marker everywhere in this code.  I see no reason to use four bytes just at this particular place.  This array contains 3-byte structures, so you cannot align them.

File: src/coreclr/md/runtime/metamodel.cpp [hunk 2]
```diff
@@ -717,7 +717,17 @@ CMiniMdBase::InitColsForTable(
...
+
+    _ASSERTE(MaxCols >= pTable->m_cCols);
+    //
+    // Mark the array of columns as not allocated (eg, not ALLOCATED_MEMORY_MARKER)
```
Comment (AntonLapounov) [inline src/coreclr/md/runtime/metamodel.cpp:726]: E.g. means “for example”. Did you actually mean "i.e."?

File: src/coreclr/md/runtime/metamodel.cpp (hunk 1 above)
Comment (AntonLapounov) [reply src/coreclr/md/runtime/metamodel.cpp:722]: Here is similar code in another function: https://github.com/dotnet/runtime/blob/2201016c1e13bdb9abf49e2e38cadf4ee0568df2/src/coreclr/md/enc/metamodelrw.cpp#L3635-L3638

Comment (AntonLapounov) [reply src/coreclr/md/runtime/metamodel.cpp:722]: And here: https://github.com/dotnet/runtime/blob/4710133ea0ed60c96ea77e66aee461eadc6c6f78/src/coreclr/md/runtime/metamodel.cpp#L880-L889

**Comment by tommcdon** [comment]
Hi @RobertHenry6bev, I've moved this PR to draft mode.  Once it is ready for review, please feel free to move out of draft mode.  Thanks for your contributions!

**Pull request closed without merging**